File: apps/mobile/src/services/matching.ts
import { api } from './api';
import { ENDPOINTS } from '../constants/api';

export interface MatchingUser {
  id: string;
  nickname: string;
  studentId: string;
  dormName: string;
  matchRate: number;
  tags: string[];
}

export interface MatchingListResponse {
  data: MatchingUser[];
  total: number;
  page: number;
  limit: number;
}

export interface ComparisonItem {
  me: boolean | number | string;
  target: boolean | number | string;
  match: boolean;
}

export interface MatchingDetailResponse {
  id: string;
  nickname: string;
  studentId: string;
  dormName: string;
  gender: string;
  nationality: string;
  matchRate: number;
  comparison: Record<string, ComparisonItem>;
  reviews: {
    id: number;
    reviewerName: string;
    content: string;
    score: number;
    createdAt: string;
  }[];
  averageReviewScore: number;
  reviewCount: number;
}

interface GetMatchingListParams {
  page?: number;
  limit?: number;
  sortBy?: 'matchRate' | 'createdAt';
  dormName?: string;
}

export async function getMatchingList(params: GetMatchingListParams = {}): Promise<MatchingListResponse> {
  const { page = 1, limit = 10, sortBy = 'matchRate', dormName } = params;

  let url = `${ENDPOINTS.MATCHING.RECOMMENDATIONS}?page=${page}&limit=${limit}&sortBy=${sortBy}`;
  if (dormName) {
    url += `&dormName=${encodeURIComponent(dormName)}`;
  }

  return api.get<MatchingListResponse>(url);
}

export async function getMatchingDetail(userId: string): Promise<MatchingDetailResponse> {
  return api.get<MatchingDetailResponse>(ENDPOINTS.MATCHING.DETAIL(userId));
}

File: apps/api/app/schemas/matching.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.user import Nationality


class MatchingUserResponse(BaseModel):
    id: str
    nickname: str | None
    studentId: int
    nationality: Nationality
    dormNames: str  # Added field
    matchRate: int
    keywords: list[str]
    isSmoker: bool
    sleepStart: int


class MatchingListResponse(BaseModel):
    total: int
    page: int
    limit: int
    data: list[MatchingUserResponse]


class ComparisonItem(BaseModel):
    me: bool | int | str
    target: bool | int | str
    match: bool


# 항목별 점수 상세 (5만원 게임 결과)
class ScoreBreakdownItem(BaseModel):
    score: int        # 유사도 점수 (0-100)
    weight: float     # 적용된 가중치
    status: str       # "Perfect", "Good", "Bad" 등


class RadarChartData(BaseModel):
    noise: int       # 소음
    clean: int       # 청결
    food: int        # 실내취식
    light: int       # 소등
    temp: int        # 온도
    time: int        # 취침시간
    habit: int       # 잠버릇


class ReviewSummary(BaseModel):
    id: int
    content: str
    score: int
    createdAt: datetime


class MatchingUserDetail(BaseModel):
    id: str
    nickname: str | None
    gender: str
    nationality: Nationality
    studentId: int
    age: int


class MatchingLifestyleDetail(BaseModel):
    dormNames: str
    isSmoker: bool
    sleepStart: int
    sleepEnd: int
    sleepHabits: str | None
    noiseLevel: int
    cleanLevel: int
    foodLevel: int
    lightLevel: int
    tempLevel: int
    homeVisit: str | None


class MatchingDetailResponse(BaseModel):
    user: MatchingUserDetail
    lifestyle: MatchingLifestyleDetail | None
    matchRate: int
    comparison: dict[str, ComparisonItem]
    radarChart: dict[str, RadarChartData]
    scoreBreakdown: dict[str, ScoreBreakdownItem]  # 항목별 점수 상세
    reviews: list[ReviewSummary]
    reviewCount: int  # Added field
    averageReviewScore: float
